salt/orch/etcd-migrate.sls

```diff
@@ -0,0 +1,155 @@
+# To migrate etcd storage version for Kubernestes execute following steps on Admin node:
+# 1. docker exec -ti $(docker ps | grep salt-master | cut -d" " -f1) salt-run state.orchestrate orch.etcd-migrate
+# 2. btrfs property set -ts /.snapshots/1/snapshot ro false
+# 3. mount -o remount,rw /
+# 4. sed -i '/etcd_version/ s/etcd2/etcd3/' /usr/share/salt/kubernetes/pillar/params.sls
+# 5. docker exec -ti $(docker ps | grep salt-master | cut -d" " -f1) salt '*' saltutil.refresh_pillar
+# 6. docker exec -ti $(docker ps | grep salt-master | cut -d" " -f1) salt '*' salt.highstate
+
+# Generic Updates
+update_pillar:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_pillar
+
+update_grains:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_grains
+
+update_mine:
+  salt.function:
+    - tgt: '*'
+    - name: mine.update
+    - require:
+       - salt: update_pillar
+       - salt: update_grains
+
+update_modules:
+  salt.function:
+    - name: saltutil.sync_modules
+    - tgt: '*'
+    - kwarg:
+        refresh: True
+    - require:
+      - salt: update_mine
+
+{%- set masters = salt.saltutil.runner('mine.get', tgt='G@roles:kube-master', fun='network.ip_addrs', tgt_type='compound') %}
+{%- for master_id in masters.keys() %}
+
+{{ master_id }}-stop-services:
+  salt.state:
+    - tgt: {{ master_id }}
+    - sls:
+      - kube-apiserver.stop
+      - kube-controller-manager.stop
+      - kube-scheduler.stop
+      - docker.stop
+      - flannel.stop
+      - etcd.stop
+
+{{ master_id }}-wait-action:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - sleep 1
+    - kwarg:
+        bg: True
+    - require:
+      - salt: {{ master_id }}-stop-services
+
+{% endfor %}
+
+{%- set workers = salt.saltutil.runner('mine.get', tgt='G@roles:kube-minion', fun='network.ip_addrs', tgt_type='compound') %}
+{%- for worker_id, ip in workers.items() %}
+
+{{ worker_id }}-stop-services:
+  salt.state:
+    - tgt: {{ worker_id }}
+    - sls:
+      - kubelet.stop
+      - kube-proxy.stop
+      - docker.stop
+      - flannel.stop
+      - etcd-proxy.stop
+
+{{ worker_id }}-wait-action:
+  salt.function:
+    - tgt: {{ worker_id }}
+    - name: cmd.run
+    - arg:
+      - sleep 1
+    - kwarg:
+        bg: True
+    - require:
+      - salt: {{ worker_id }}-stop-services
+
+{% endfor %}
+
+{%- for master_id in masters.keys() %}
+
+{{ master_id }}-backup-etcd:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - mkdir /tmp/backup; btrfs subvolume snapshot /var/lib/etcd /tmp/backup/etcd
+
+{{ master_id }}-migrate-etcd:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - set -a; source /etc/sysconfig/etcdctl; env ETCD_API=3 etcdctl migrate --data-dir=/var/lib/etcd; echo "ETCD_API=3" > /var/lib/etcd/version
```

Comment:
Is this something etcd reads, I don't see anything in the docs?

    echo "ETCD_API=3" > /var/lib/etcd/version

If it is, we should be setting it based on the `etcd_version` pillar within the etcd state itself, new machines that start life with etcd3 will never have this migration ran to create that file.

Reply:
It was only for testing